pkg/controllers/rollout/controller.go

```diff
@@ -0,0 +1,278 @@
+	}
+	// check that the crp is not being deleted
+	if crp.DeletionTimestamp != nil {
+		klog.V(2).InfoS("Ignoring clusterResourcePlacement that is being deleted", "clusterResourcePlacement", crpName)
+		return ctrl.Result{}, nil
+	}
+	// check that it's actually rollingUpdate strategy
+	if crp.Spec.Strategy.Type != fleetv1beta1.RollingUpdateRolloutStrategyType {
+		klog.V(2).InfoS("Ignoring clusterResourcePlacement with non-rolling-update strategy", "clusterResourcePlacement", crpName)
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// pickBindingsToRoll lists all bindings associated with a CRP and returns the bindings that are ready to be rolled.
+func (r *Reconciler) pickBindingsToRoll(ctx context.Context, latestResourceSnapshotName string, crp *fleetv1beta1.ClusterResourcePlacement) ([]*fleetv1beta1.ClusterResourceBinding, []*fleetv1beta1.ClusterResourceBinding, error) {
+	// Those are the bindings that are chosen by the scheduler to be applied to selected clusters.
+	// They include the bindings that are already applied to the clusters and the bindings that are newly selected by the scheduler.
+	schedulerTargetedBinds := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// The content of those bindings that are considered to be already running on the targeted clusters.
+	readyBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that have the potential to be ready during the rolling phase.
+	// It includes all bindings that have been applied to the clusters and not deleted yet so that they can still be ready at any time.
+	canBeReadyBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that have the potential to be unavailable during the rolling phase which
+	// includes the bindings that are being deleted. It depends on work generator and member agent for the timing of the removal from the cluster.
+	canBeUnavailableBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that are candidates to be updated during the rolling phase.
+	updateCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+	// the subset of updateCandidates that we pick to update
+	tobeUpgradedBinding := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that are candidates to be removed during the rolling phase.
+	removeCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+	// the subset of removeCandidates that we pick to remove
+	tobeRemovedBinding := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// calculate the cutoff time for a binding to be applied before so that it can be considered ready
+	var unavailablePeriod time.Duration
+	if crp.Spec.Strategy.RollingUpdate.UnavailablePeriodSeconds != nil {
+		unavailablePeriod = time.Duration(*crp.Spec.Strategy.RollingUpdate.UnavailablePeriodSeconds)
+	} else {
+		unavailablePeriod = 60 * time.Second
```

Comment:
Hi Ryan! It might be better to make this default value a top-level constant IMO.

Reply:
this is the CRD default value, will ETCD/API server honor the CRE schema?